src/Common/src/Interop/Windows/winhttp/Interop.SafeWinHttpHandle.cs

```diff
@@ -11,30 +12,51 @@ internal partial class Interop
     internal partial class WinHttp
     {
         // Issue 2501: Fold SafeWinHttpHandleWithCallback and SafeWinHttpHandle into a single type.
-        //
-        // SafeWinHttpHandleWithCallback is a better pattern; SafeWinHttpHandle is probably a better
-        // name. The implementation for the former should probably be folded into the latter.
-
+        // SafeWinHttpHandleWithCallback is incorrectly overriding the Dispose(bool disposing) method
+        // and will be fixed as part of merging the classes.
         internal class SafeWinHttpHandle : SafeHandleZeroOrMinusOneIsInvalid
         {
+            private SafeWinHttpHandle _parentHandle = null;
+            
             public SafeWinHttpHandle() : base(true)
             {
             }
 
-            public static void DisposeAndClearHandle(ref SafeWinHttpHandle winHttpHandler)
+            public static void DisposeAndClearHandle(ref SafeWinHttpHandle safeHandle)
             {
-                if (winHttpHandler != null)
+                if (safeHandle != null)
                 {
-                    winHttpHandler.Dispose();
-                    winHttpHandler = null;
+                    safeHandle.Dispose();
+                    safeHandle = null;
                 }
             }
 
+            public void SetParentHandle(SafeWinHttpHandle parentHandle)
+            {
+                Debug.Assert(_parentHandle == null);
+                Debug.Assert(parentHandle != null);
+                Debug.Assert(!parentHandle.IsInvalid);
+
+                bool success = false;
+                parentHandle.DangerousAddRef(ref success);
+                if (!success)
```

Comment:
This check isn't necessary: DangerousAddRef will throw if it's not successful.  The reason it has the ref argument as well is so that it can be used reliably in try/finally blocks, but that obviously doesn't apply here.